### ISSUE TITLE
http OPTIONS call returns 500 error on 0.37.2

### ISSUE THREAD

**Issue by d3ming** [Miserlou/Zappa#699]
## Context
Python 2.7, virtualenv

I'm using Flask with Zappa, and after upgrading to 0.37.2 (from 0.36.0) the OPTIONS call to my API will return a 500 'Internal server error'.  The error is from Cloudfront, and I don't even see the request hitting my server.

When I reverted to 0.36.0 and redeploy my API, it works again.  It also works locally, if I call my API against localhost from starting flask locally.  So this appears to be a regression from 0.37+

## Expected Behavior
OPTIONS call against an endpoint should return 200

## Actual Behavior
500, internal server error from cloudfront

## Possible Fix
Reverting back to using 0.36.0 and redeploy

## Steps to Reproduce
0. Have an Flask API endpoint deployed with Zappa 0.36.0, at say `http://my.api.com/dev/api/test`
1. Upgrade zappa to 0.37.2
2. zappa update dev
3. `OPTIONS http://my.api.com/dev/api/test`

## Your Environment
- Zappa: 0.37.2
- Python: 2.7 with virtualenv
- OSX El Capitan (10.11.6)

**Comment by Miserlou** [comment]
Dang, was hoping this was fixed.

Can you tell me what's in your OPTIONS body? Is there nothing? 

We need to specifically handle the OPTIONS call in both body and non-body scenarios. See: https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol#Summary_table

I suppose we should add a line to check if there is a body there, I imagine that's the error? Can you give me the `zappa tail`?

Related: https://github.com/Miserlou/Zappa/issues/696

**Comment by Miserlou** [comment]
Do you have `cors: true` in your `zappa_settings`?

**Comment by d3ming** [comment]
Yup, I do have `cors: true` set.  And the OPTIONS response body is just empty

**Comment by Miserlou** [comment]
Removing this key fixes the 500 for me.

Tell me - do you actually use the Zappa/Amazon provided CORS features? Should this feature be deprecated in favor of application-layer CORS?

Pinging @collingreen @mathom

**Comment by Miserlou** [comment]
You can leave `cors: true` but set `binary_support: false` - they don't appear to be playing nicely together.

**Comment by mathom** [comment]
Can't say I use it myself. App level CORS seems more useful for a project like Zappa. Just my 2 cents though.

**Comment by d3ming** [comment]
Turning `cors` off worked, and yeah apparently I did not need to use this setting.

**Comment by abdulwahid24** [comment]
@d3ming I was also facing the same issue and it got resolve by setting up the `binary_search` to `false` but at the same time it raise an issue with image upload, So in this I need to have `binary_search` to `true`. 
Let me know what should I do in this situation as I am having cors enable at django side for my rest api and using these api in angular js for hybrid mobile app. Though I tried to disable the preflight option request in angularjs but its really hard and bad hack to do it as its automatically get added by chrome.

Please let me know how should I deal with this situation.

Thanks.

**Comment by Miserlou** [comment]
You have two avenues: a) try to fix simultaneous CORS + Binary Support in Zappa itself, thereby fixing your problem and everybody else's or b) disable `cors` and then add in your own CORS support at your own application layer. 

I strongly suggest a!

**Comment by abdulwahid24** [comment]
Hey @Miserlou, You are absolutely right, as of now I used option `b` and It would be my honor to contribute to zappa so I am would like to implement the feature `a`  and would raise PR soon. 
Thanks you very much. 👍

**Comment by wobeng** [comment]
Any update on this?

**Comment by eely22** [comment]
This was one of the issues affecting us, so I dug around trying to find the solution. I found an answer on an unrelated project here: https://github.com/awslabs/aws-serverless-express/issues/58

I tried that out, adding the contentHandling property to the OPTIONS methods, and it seems to fix the problem. This is the first time I have had to dig this far into API Gateway, so if anyone sees or knows any issue with this approach, let me know, but it seems to solve it in all the cases I tried.

**Comment by IlyaSukhanov** [comment]
"Tell me - do you actually use the Zappa/Amazon provided CORS features? Should this feature be deprecated in favor of application-layer CORS?"

I don't see an option of using application level `CORS` with `API Key`. When `API Key` is set all requests, including `OPTIONS` return `403` when `x-api-key` key is omitted. Naturally browser will not supply `x-api-key` when performing CORS checks.

Although I do think `API Key` being incompatible with application-layer CORS is a separate issue, @Miserlou do you agree? If so I'll open a new one.

**Comment by eely22** [comment]
I completely agree with @IlyaSukhanov, we need the cors feature for this very reason as we also use AWS API Keys.
